pkg/controller/gcppubsub/reconcile.go

```diff
@@ -0,0 +1,328 @@
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Insert(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *reconciler) removeFinalizer(s *v1alpha1.GcpPubSubSource) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Delete(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *reconciler) createReceiveAdapter(ctx context.Context, src *v1alpha1.GcpPubSubSource, subscriptionId, sinkURI string) (*v1.Deployment, error) {
+	ra, err := r.getReceiveAdapter(ctx, src)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get an existing receive adapter", zap.Error(err))
+		return nil, err
+	}
+	if ra != nil {
+		logging.FromContext(ctx).Desugar().Info("Reusing existing receive adapter", zap.Any("receiveAdapter", ra))
+		return ra, nil
+	}
+	svc := r.makeReceiveAdapter(src, subscriptionId, sinkURI)
+	err = r.client.Create(ctx, svc)
+	logging.FromContext(ctx).Desugar().Info("Receive Adapter created.", zap.Error(err), zap.Any("receiveAdapter", svc))
+	return svc, err
+}
+
+func (r *reconciler) getReceiveAdapter(ctx context.Context, src *v1alpha1.GcpPubSubSource) (*v1.Deployment, error) {
+	sl := &v1.DeploymentList{}
+	err := r.client.List(ctx, &client.ListOptions{
+		Namespace:     src.Namespace,
+		LabelSelector: r.getLabelSelector(src),
+		// TODO this is only needed by the fake client. Real K8s does not need it. Remove it once
+		// the fake is fixed.
+		Raw: &metav1.ListOptions{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Deployment",
+			},
+		},
+	},
+		sl)
+
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Unable to list deployments: %v", zap.Error(err))
+		return nil, err
+	}
+	for _, svc := range sl.Items {
+		if metav1.IsControlledBy(&svc, src) {
+			return &svc, nil
+		}
+	}
+	return nil, apierrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+func (r *reconciler) getLabelSelector(src *v1alpha1.GcpPubSubSource) labels.Selector {
+	ls := labels.NewSelector()
+	for k, v := range getLabels(src) {
+		req, err := labels.NewRequirement(k, selection.Equals, []string{v})
+		if err != nil {
+			// Probably fail.
+		}
+		ls.Add(*req)
+	}
+	return ls
+}
+
+func getLabels(src *v1alpha1.GcpPubSubSource) map[string]string {
+	return map[string]string{
+		"knative-eventing-source":      controllerAgentName,
+		"knative-eventing-source-name": src.Name,
+	}
+}
+
+func (r *reconciler) makeReceiveAdapter(src *v1alpha1.GcpPubSubSource, subscriptionId, sinkURI string) *v1.Deployment {
+	credsVolume := "google-cloud-key"
+	credsMountPath := "/var/secrets/google"
+	credsFile := fmt.Sprintf("%s/key.json", credsMountPath)
+	replicas := int32(1)
+	dLabels := getLabels(src)
+	return &v1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    src.Namespace,
+			GenerateName: fmt.Sprintf("gcppubsub-%s-", src.Name),
+			Labels:       dLabels,
+			OwnerReferences: []metav1.OwnerReference{
```

Comment:
Why not use the util function for setting the owner reference? https://github.com/knative/eventing-sources/blob/master/pkg/controller/kuberneteseventsource/reconcile.go#L105

Reply:
Done.